### ISSUE TITLE
ALL/ANY/SOME do not accept Value Lists.

### ISSUE THREAD

**Issue by petersonschmitt** [JSQLParser/JSqlParser#1232]
**Describe the bug**
Error when converting sql using "CCJSqlParserUtil.parse(sql)", sql with condition "nin ( != )" others like `eq / neq / in / isnull and isnnull` work normal
**To Reproduce**
1. Example SQL
   `select t0.total_value, t0.code, t0.expected_at, t0.description, t0.id, t0.status, t5.id, t5.name, t5.id, t4.id, t4.id, t1.id, t1 .name, t2.id, t2.name, t3.id, t3.description from opportunity t0 left join "user" t5 on t5.id = t0.user_id left join stage t4 on t4.id = t0.stage_id left join customer t1 on t1.id = t0.customer_id left join lead t2 on t2.id = t0.lead_id left join opportunity_type t3 on t3.id = t0.opportunity_type_id where t0.pipeline_id = ? and t0.status = ? and t0.updated_at > ? and (t0.customer_id != all(?::uuid[]) or t0.customer_id is null ) order by t0.total_value desc, t0.code desc`
2. Parsing this SQL using: 
     `CCJSqlParserUtil.parse(sql)`
3. Exception
```
Parameters: {rankBy=[stage.id],q=[{pipeline.id->eq->4542a6a0-29a5-4bd8-9857-314f20e648a3}{status->eq->PROGRESS}{updatedAt->gt->2021-03-04T02:59:59.999Z}({customer.id->nin->[1e7d13d3-b43a-4651-a3ca-0b78a5fa84ee]}||{customer.id->isnull}){orderby->[totalValue desc,code desc]}{select->[id,description,code,status,totalValue,expectedAt,lead.name,customer.name,opportunityType.description,user.id,user.name,stage.id,activityCount]}{pageSize->10}{firstRow->0}]}
java.lang.IllegalArgumentException: net.sf.jsqlparser.JSQLParserException: Encountered unexpected token: "!=" <OP_NOTEQUALSBANG>
    at line 1, column 523.

Was expecting one of:

    "&"
    "&&"
    ")"
    "::"
    "<<"
    ">>"
    "AND"
    "COLLATE"
    "OR"
    "["
    "^"
    "|"

	at com.ws.opportunity.opportunity.OpportunityRankQueryFactory.build(OpportunityRankQueryFactory.java:74) ~[main/:?]
	at com.ws.opportunity.opportunity.OpportunityRankQueryFactory$Proxy$_$$_WeldClientProxy.build(Unknown Source) ~[main/:?]
	at com.ws.opportunity.opportunity.OpportunityDAO.list(OpportunityDAO.java:195) ~[main/:?]
	at com.ws.opportunity.opportunity.OpportunityService.findList(OpportunityService.java:159) ~[main/:?]
	at com.ws.opportunity.opportunity.OpportunityResource.list(OpportunityResource.java:128) ~[main/:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_281]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_281]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_281]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_281]
```

**System**
Database -> Postgres
Java Version-> Java 8
JSqlParser version -> tried with 3.2 and 4.0

**Comment by manticore-projects** [comment]
Greetings.

Actually, it is not about the `!=`, this modified statement parses just fine:
```sql
SELECT  t0.total_value
        , t0.code
        , t0.expected_at
        , t0.description
        , t0.id
        , t0.status
        , t5.id
        , t5.name
        , t5.id
        , t4.id
        , t4.id
        , t1.id
        , t1.name
        , t2.id
        , t2.name
        , t3.id
        , t3.description
FROM opportunity t0
    LEFT JOIN "user" t5
        ON t5.id = t0.user_id
    LEFT JOIN stage t4
        ON t4.id = t0.stage_id
    LEFT JOIN customer t1
        ON t1.id = t0.customer_id
    LEFT JOIN lead t2
        ON t2.id = t0.lead_id
    LEFT JOIN opportunity_type t3
        ON t3.id = t0.opportunity_type_id
WHERE t0.pipeline_id = ?
    AND t0.status = ?
    AND t0.updated_at > ?
    AND ( t0.customer_id != 3 /* all(?::uuid[]) */
            OR t0.customer_id IS NULL )
ORDER BY    t0.total_value DESC
            , t0.code DESC
;
```

The issue is about the `all(?::uuid[])`. You can test it [here online and interactively](http://jsqlformatter.manticore-projects.com/jsqlformatter/demo.html?args=-c%20MoUQMiDCAqAEsBcAMA6BB7BBDANgfQDdcBXAUwCh4qqAaRVAY3QBMLrq7kVSAPAB1IMEpZniwJK7eJ1SsAzgwBOASz4Jl6AHaSpMlMuY72eudgTE5RjogCs+w1Om2UmrAFs2jznYNXaiABZ7PycEIN9HWE4ARmDImJd3T11EACY4rzTEjxCoxABmDJSEQvklVXUtcgAxACUAeQBZWHQ+PnRFc01lBABPeiMIargAKXqASQA5WAAiC1JFGdtc+umEH2ZYAF56FHnFPAiqIdGJ6dMsAHNSQJW18M2drgvrw4djkGHYManYBgsMB5FIhoncQfZtrt-qZ0EC3oNPqdfjhSFhNghUmCMRCnqgUWj4dQTt8zi02h0uj1eng+gICljCgZIVxWu1OsRun0ab0BPCAOoACRAtRAuz4qlIOGUmlIb0hAH4jABBSYAEV2F3McgVyrVu2IfGY4hEYgQsAAfLBFdQVeqABRQgGwhZygCEO1wODt8oAXD7iMQDABtAC6AEpctR6rVHTC4UzxsBYJMAKpgMCwCPR1XC2AAIQAmtQuBhsPgiDgyLAc8BIJH-FwmKxqyBa+QANxAA+)

What does `all(?::uuid[])` doe exactly? Can you point me onto the reference API please?

**Comment by manticore-projects** [comment]
I assume, that `all` is a reserved keyword and for that reason is not allowed to be used as a function here.

The parser would accept:
```sql
t0.customer_id != to_char(?::uid[])
```

**Comment by manticore-projects** [comment]
There is another collision:
```
Expression AllComparisonExpression() :
{
    AllComparisonExpression retval = null;
    SubSelect subselect = null;
}
{
 <K_ALL> "(" subselect=SubSelect() ")" { retval = new AllComparisonExpression(subselect); }
   {
      return retval;
   }
}
```

**Comment by manticore-projects** [comment]
@Wumpz: Please I would like to ask for your feedback.
```
Expression AnyComparisonExpression() :
{
    AnyComparisonExpression retval = null;
    AnyType anyType;
    SubSelect subselect = null;
}
{
 (<K_ANY> { anyType = AnyType.ANY; } | <K_SOME> { anyType = AnyType.SOME; } )
 "(" subselect=SubSelect() ")" { retval = new AnyComparisonExpression(anyType, subselect); }
   {
      return retval;
   }
}
```

`ANY` Expression supports a Type [`ANY`, `SOME`], but for any reason `ALL` is a separate Expression. Should we not merge this, when incorporating the ValueLists?

**Comment by manticore-projects** [comment]
Quick test using H2:
```sql
-- Valid and supported by JSQLParser
select *
from dual
where 1 = ALL( select 1 from dual );

-- Valid but not yet supported by JSQLParser
select *
from dual
where 1 = ALL( VALUES 1,2,3 );

-- Invalid
select *
from dual
where 1 = ALL( VALUES(1,2,3) );
```

**Comment by manticore-projects** [comment]
@petersonschmitt: You can make your statement work now already  by using a proper subquery
```sql
SELECT  t0.total_value
        , t0.code
        , t0.expected_at
        , t0.description
        , t0.id
        , t0.status
        , t5.id
        , t5.name
        , t5.id
        , t4.id
        , t4.id
        , t1.id
        , t1.name
        , t2.id
        , t2.name
        , t3.id
        , t3.description
FROM opportunity t0
    LEFT JOIN "user" t5
        ON t5.id = t0.user_id
    LEFT JOIN stage t4
        ON t4.id = t0.stage_id
    LEFT JOIN customer t1
        ON t1.id = t0.customer_id
    LEFT JOIN lead t2
        ON t2.id = t0.lead_id
    LEFT JOIN opportunity_type t3
        ON t3.id = t0.opportunity_type_id
WHERE t0.pipeline_id = ?
    AND t0.status = ?
    AND t0.updated_at > ?
    AND ( t0.customer_id != ALL (SELECT ?::uuid[])
            OR t0.customer_id IS NULL )
ORDER BY    t0.total_value DESC
            , t0.code DESC
;
```
(Test it online here)[http://jsqlformatter.manticore-projects.com/jsqlformatter/demo.html?args=-c%20MoUQMiDCAqAEsBcAMA6BB7BBDANgfQDdcBXAUwCh4qqAaRVAY3QBMLrq7kVSAPAB1IMEpZniwJK7eJ1SsAzgwBOASz4Jl6AHaSpMlMuY72eudgTE5RjogCs+w1Om2UmrAFs2jznYNXaiABZ7PycEIN9HWE4ARmDImJd3T11EACY4rzTEjxCoxABmDJSEQvklVXUtcgAxACUAeQBZWHQ+PnRFc01lBABPeiMIargAKXqASQA5WAAiC1JFGdtc+umEH2ZYAF56FHnFPAiqIdGJ6dMsAHNSQJW18M2drgvrw4djkGHYManYBgsMB5FIhoncQfZtrt-qZ0EC3oNPqdfjhSFhNghUmCMRCnqgUWj4dQTt8zi02h0uj1eng+gICljCgZIVxWu1OsRun0ab0BPCAOoACRAtRAuz4qlIOGUmlIb0hAH4jABBSYAEV2F3McgVyrVu2IfGY4hEYgQsAAfLBFdQVeqABRQgGwhZygCEOyVYDAsDtoAgMCtAC5A8RiAYANoAXQAlLlqPVao6YXCmeNgLBJgBVL2wWMJ1XC2AAIQAmtQuBhsPgiDgyLAC8BIHH-FwmKx6yBG+QANxAA+]

**Comment by manticore-projects** [comment]
Maybe change the Title of the issue into something like `ALL/ANY/SOME do not accept Value Lists`.

**Comment by manticore-projects** [comment]
Solved by PR #1233
